src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/AbstractEvaluationCommandLineHandler.cs

```diff
@@ -0,0 +1,243 @@
+            Requires.NotNull(metadata, nameof(metadata));
+
+            if (!difference.AnyChanges)
+                return;
+
+            difference = NormalizeDifferences(difference);
+            EnqueueEvaluation(version, difference);
+
+            ApplyChangesToContext(version, difference, metadata, isActiveContext);
+        }
+
+        /// <summary>
+        ///     Applys the specified version of design-time build <see cref="IProjectChangeDiff"/> to the underlying
+        ///     <see cref="IWorkspaceProjectContext"/>, indicating if the context is the currently active one.
+        /// </summary>
+        /// <exception cref="ArgumentNullException">
+        ///     <paramref name="version"/> is <see langword="null"/>.
+        ///     <para>
+        ///         -or-
+        ///     </para>
+        ///     <paramref name="difference" /> is <see langword="null"/>.
+        /// </exception>
+        public void ApplyDesignTimeChanges(IComparable version, IProjectChangeDiff difference, bool isActiveContext)
+        {
+            Requires.NotNull(version, nameof(version));
+            Requires.NotNull(difference, nameof(difference));
+
+            if (!difference.AnyChanges)
+                return;
+
+            difference = NormalizeDifferences(difference);
+            difference = ResolveDesignTimeConflicts(version, difference);
+
+            ApplyChangesToContext(version, difference, ImmutableDictionary<string, IImmutableDictionary<string, string>>.Empty, isActiveContext);
+        }
+
+        protected abstract void AddToContext(string fullPath, IImmutableDictionary<string, string> metadata, bool isActiveContext);
+
+        protected abstract void RemoveFromContext(string fullPath);
+
+        private void ApplyChangesToContext(IComparable version, IProjectChangeDiff difference, IImmutableDictionary<string, IImmutableDictionary<string, string>> metadata, bool isActiveContext)
+        {
+            foreach (string includePath in difference.RemovedItems)
+            {
+                RemoveFromContextIfPresent(includePath);
+            }
+
+            foreach (string includePath in difference.AddedItems)
+            {
+                AddToContextIfNotPresent(includePath, metadata, isActiveContext);
+            }
+
+            // We Remove then Add changed items to pick up the Linked metadata
+            foreach (string includePath in difference.ChangedItems)
+            {
+                RemoveFromContextIfPresent(includePath);
+                AddToContextIfNotPresent(includePath, metadata, isActiveContext);
+            }
+
+            Assumes.True(difference.RenamedItems.Count == 0, "We should have normalized renames.");
+        }
+
+        private void RemoveFromContextIfPresent(string includePath)
+        {
+            string fullPath = _project.MakeRooted(includePath);
+            if (_paths.Remove(fullPath))
+            {
+                RemoveFromContext(fullPath);
```

Comment:
Theoretically it doesn't fail, however, looking into this right now trying out Roslyn solution and looks like Roslyn does throw in situations I didn't expect (such as duplicate references). 

This behavior we have right now is the same as we've always had, but agreed I think i want to avoid messing up the bookkeeping even in the case of bugs.